### ISSUE TITLE
Basic architectural design

### ISSUE THREAD

**Issue by rschroll** [jupyterlab/jupyterlab_app#2]
Following up on our video chat discussion: We need to determine the basic architecture for the desktop application.  There are (at least) two general approaches:

1. A thin application that offers a browser to view a JupyterLab server running locally or remotely.  This is the approach taken by [Juno](https://github.com/rschroll/juno).  Desktop integration of things like menus and dialogs could be achieved by injecting JavaScript into the browser window, either to monkey patch some features or to connect to hooks provided by JupyterLab.

Pros:
- Smaller code base.
- Easy support for both local and remote servers.

Cons:
- Introduces another interface, which will presumably need to be versioned.

2. A thick application that ships with some / most / all of the JupyterLab browser-side code.  Relevant UI modules could be easily swapped out for desktop replacements.  The app would load this content, and then instruct it to contact a server to complete the setup.  (I'm doing my best to summarize @ellisonbg's description, but I may be screwing it up a bit.)

Pros:
- Easier insertion of desktop customizations.
- Split may occur along existing interface (?)
- Potential to build easy system for installing additional extensions.

Cons:
- Interface likely more extensive, may be more sensitive to versioning.
- May not be easy to serve JupyterLab resource without Jupyter server.  (Based on my 10 minute exploration of the code; I'll defer to others on this point.)

Personally, I'm prejudiced in favor of (1).  Part of that is that I've worked in that direction, and things seem to be working.  Part of it is that it will probably lead to a working app faster.  (Arguably, it's already there.)  And part of it is that I don't understand the JupyterLab architecture enough to really judge the amount of work required for (2).

What say all of you?  Is my understanding mostly right?  Am I missing another architectural approach?

**Comment by ddavidebor** [comment]
I was looking into this a couple of months ago with my colleagues and were going in the direction of a hybrid approach:

* A complete vanilla jupyterlab server
* An Electron-based UI with all the javascript and static content embedded

The Electron app would initialize the Jupyterlab UI, completely ignoring the javascript from the Jupiter server, and maybe just extracting the necessary info from the HTML in the main page.

Basically retaining the Jupyter server vanilla while shipping a modified UI served locally, regardless of where the server is.

Where can i find @ellisonbg description?

**Comment by ellisonbg** [comment]
Some comments on implementation questions...

* The JupyterLab uses a stock vanilla jupyter server, version 5.0 and above. The only modifications needed are the server extension for serving up the JLab UI/JS stuff.
* Both of these approaches can support both remote and local servers without any problems.
* While option 2) would require more code, most of it is in standalone npm packages that already exist in JupyterLab and can be reused.
* The work for option 2 is pretty simple, given our current architecture.
* All of the logic for starting and connecting to servers will be exactly the same for option 1/2.

**Comment by ddavidebor** [comment]
From the last chat:
  * use jupyter file management, not nodejs native one

**Comment by jasongrout** [comment]
Fyi, I thought we supported notebook 4.4 as well.

**Comment by jasongrout** [comment]
Or at least I thought it worked on notebook 4.4, even if we don't officially support it.

**Comment by rschroll** [comment]
@ellisonbg: If we can run JupyterLab against a stock Jupyter server, that's very exciting, and I'd think it's the way to go.  It opens up the possibility of running this app against cloud providers like tmpnb and azure notebooks, which seems pretty cool to me.

My understanding is that JupyterLab currently uses the existing notebook server to serve up its HTML/JS/CSS. (Please correct me if I'm wrong.) I wonder if an interesting first step would be to write a little server that serves just the JupyterLab resources, and points to another server to provide the stock Jupyter resources.  This could use a standard browser, so we don't have to worry about the electron side right now.  (I wonder about cross-site scripting issues, but for testing purposes, we can probably just launch the browser with the flags to disable these sorts of checks.)  Once that's working, we can either move the server entirely into the electron app, or just assemble the static resources from the server into the electron app's www/ folder.

**Comment by ddavidebor** [comment]
> I wonder if an interesting first step would be to write a little server that serves just the JupyterLab resources, and points to another server to provide the stock Jupyter resources. 

It's even easier that that. With the electron framework you can (and normally do) serve resources directly from disk. The same way you can open a local .html file with the browser. No need for extra local http servers.

**Comment by blink1073** [comment]
JupyterLab requires Notebook 4.3+.

**Comment by ddavidebor** [comment]
Now the approach is well delineated in the code, i think we can close this issue to avoid clutter.
